### PR TITLE
Include test files when not linked in setup.py

### PR DESCRIPTION
**Pull request by Unknwon**
Addresses #49

### CODE REVIEW

**Comment by beyang** [comment]
Can you add a test case?

**Comment by beyang** [comment]
Has this been tested on Flask and other popular Python repositories?

File: grapher/scan.py [hunk 1]
```diff
@@ -72,13 +72,28 @@ def source_files_for_pip_unit(unit_dir: str) -> List[str]:
...
         for pkg_file in pkg_files:
-            files.append(normalize(os.path.join(pkg.replace('.', '/'), pkg_file)))
+            files.append(normalize(os.path.join(pkg_path, pkg_file)))
+
+    # Make good guess for test files when they are not linked.
+    if not included_tests:
+        for cand in tests_candidates:
+            pkg_files = get_source_files(cand)
+            for pkg_file in pkg_files:
+                files.append(normalize(os.path.join(cand, pkg_file)))
```
Comment (beyang) [inline grapher/scan.py]: Initially, I was thinking of having the tests be a part of a separate source unit. Many times in Python, the tests are in a completely separate directory tree and use the library as an external API consumer. Is there a compelling reason to have them as part of the same source unit?

**Comment by beyang** [comment]
Can you add a test case to test for this?

File: grapher/scan.py [hunk 2]
```diff
@@ -62,7 +62,9 @@ def find_pip_pkgs(rootdir: str) -> List:
             setup_dict_to_json_serializable_dict(setup_dict, rootdir=os.path.relpath(setup_dir, rootdir)))
     return setup_infos
 
-def source_files_for_pip_unit(unit_dir: str) -> List[str]:
+TEST_DIR = "tests"
```
Comment (beyang) [inline grapher/scan.py]: Add docstring.

File: grapher/scan.py [hunk 3]
```diff
@@ -72,28 +74,45 @@ def source_files_for_pip_unit(unit_dir: str) -> List[str]:
...
+    return files, tests
 
-# pkgToUnit transforms a Pip package struct into a source unit.
-def pkgToUnit(pkg: Dict) -> Unit:
+# pkgToUnit transforms a Pip package struct into a list of source units,
+# including main unit and possible test unit.
+def pkgToUnit(pkg: Dict) -> List[Unit]:
     pkgdir = pkg['rootdir']
-    files = source_files_for_pip_unit(pkgdir)
+    files, tests = source_files_for_pip_unit(pkgdir)
```
Comment (beyang) [inline grapher/scan.py]: `s/tests/test_files/` < "test_files" is a better name.

File: grapher/scan.py [hunk 4]
```diff
@@ -106,6 +125,23 @@ def pkgToUnit(pkg: Dict) -> Unit:
...
+    if len(tests) == 0:
+        return [unit]
+    return [unit, Unit(
+        Name = pkg['project_name'],
+        Type = TEST_UNIT_KEY.Type,
+        Repo = "",
+        CommitID = "",
+        Files = sorted(tests),
+        Dir = TEST_DIR,
+        Dependencies = [UnitKey(
```
Comment (beyang) [inline grapher/scan.py]: can you use `unit.key()` here or something like
```
Name = unit.Name,
Type = unit.Type,
Repo = unit.Repo,
CommitID = unit.CommitID,
Version = unit.Version,
```

That way, if anything changes to unit, you won't also have to update this code.

**Comment by beyang** [comment]
LGTM after comments addressed.

**Comment by beyang** [comment]
Can you add some code to the test file that references code in the main library? This is a common pattern and we want to test for the fact that these references resolve properly.

**Comment by Unknwon** [comment]
@beyang I can't seem to figure out how to properly import main lib as a module in test file...

Another thing is `srclib test` does not seem to generate `*.depresolve.json` and `*.graph.json` for test source unit, only `TestFile.unit.json` is generated (see failed CI).

**Comment by beyang** [comment]
@Unknwon messaged you on Slack

**Comment by beyang** [comment]
The test->main-lib reference is incorrect.

You have `pkg0.foo()` here: https://github.com/sgtest/python-link-tests/blob/9f49c24191d003fba11e5f045b1cc6763aaf92b0/pkg0/tests/test_m0.py#L7

But in the `setup.py`, you have `py_modules=['m0']`, which means when this package is installed, it should be referenced by doing `import m0`. Based on the `setup.py`, a user of the main library has no idea about `pkg0` and the test package is just another user of the main library.

**Comment by beyang** [comment]
Rename `TestFile` to `PythonTtestPackage` (a more accurate description as it specifies the language of the source unit and "package" is a better term because most people don't equate "file" and "directory").

**Pull request merged**